### PR TITLE
Fix build issues detected by newer tools

### PR DESCRIPTION
**Pull request by sjlongland**
Changes in dependencies and test tools triggered a build failure in the `main` branch regarding the `Metadata` data type.

Furthermore, a stricter `pip` caused issues with doing documentation builds.

This corrects both issues: test-merged with #184 in `nomerge/metadata-and-precommit-changes` to confirm bugfix.

### CODE REVIEW

**Comment by sjlongland** [comment]
Of course, `precommit` doesn't run without a PR… one moment…

**Comment by sjlongland** [comment]
Not identical, but this seems eerily similar: https://github.com/python/mypy/issues/14602

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -42,7 +42,9 @@ jobs:
...
       - name: Setup Python
         uses: actions/setup-python@v4
+      - name: Set up venv
+        run: python -m venv .pyenv
       - name: Install tox
-        run: python -m pip install tox
-      - name: Run
-        run: tox -e docs
+        run: .pyenv/bin/python -m pip install tox --break-system-packages
+      - name: Run tox
```
Comment (diazona) [inline .github/workflows/docs.yml:49]: I think rather than using `--break-system-packages`, I'd prefer to either pass `--user` (if that works), or install `pipx` if necessary and then use that to install tox (although it probably still requires `--user` to install `pipx`, so that doesn't really gain anything), or use the virtualenv from the previous step to install tox; or maybe it's time to switch to [the official tox GitHub Action](https://github.com/tox-dev/tox-gh). The setup for the official action seems somewhat more complicated, but it might be worth it in the long run.

**Review by diazona — CHANGES_REQUESTED** [review]

File: .github/workflows/docs.yml [hunk 2]
```diff
@@ -42,7 +42,9 @@ jobs:
...
+        run: .pyenv/bin/python -m tox -e docs
```
[the remark this answers is not in the document]
Comment (sjlongland) [reply .github/workflows/docs.yml]: Yeah, I meant to remove that `--break-system-packages` but obviously didn't, it's gone now.

File: test_support/test_support/metadata.py [hunk 3]
```diff
@@ -271,8 +271,10 @@ def is_at_least(required: str, *, v=metadata_version):
         dependencies,
         optional_dependencies,
         entry_points,
-        authors,
-        maintainers,
+        # mypy complains that List[Tuple[str, str]] doesn't match
+        # List[Tuple[str, Optional[str]]], this seems like a bug.
+        authors,  # type: ignore[arg-type]
+        maintainers,  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (sjlongland) [reply test_support/test_support/metadata.py:277]: Yeah, using `typing.cast` didn't help:

```
test_support/test_support/metadata.py:274: error: Argument 10 to "StandardMetadata" has incompatible type "list[tuple[str, str | None]]"; expected "list[tuple[str, str]]"  [arg-type]
test_support/test_support/metadata.py:275: error: Argument 11 to "StandardMetadata" has incompatible type "list[tuple[str, str | None]]"; expected "list[tuple[str, str]]"  [arg-type]
Found 2 errors in 1 file (checked 1 source file)
```

[the remark this answers is not in the document]
Comment (sjlongland) [reply test_support/test_support/metadata.py]: and if I remove it, I get this:
```
test_support/test_support/metadata.py:274: error: Argument 10 to "StandardMetadata" has incompatible type "list[tuple[str, str]]"; expected "list[tuple[str, str | None]]"  [arg-type]
test_support/test_support/metadata.py:274: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
test_support/test_support/metadata.py:274: note: Consider using "Sequence" instead, which is covariant
test_support/test_support/metadata.py:275: error: Argument 11 to "StandardMetadata" has incompatible type "list[tuple[str, str]]"; expected "list[tuple[str, str | None]]"  [arg-type]
test_support/test_support/metadata.py:275: note: "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
test_support/test_support/metadata.py:275: note: Consider using "Sequence" instead, which is covariant
```

It's thoroughly confused.

[the remark this answers is not in the document]
Comment (diazona) [reply test_support/test_support/metadata.py]: Hmm, so I did some local testing and I had it backwards: it's complaining about passing a value of type `List[Tuple[str, str]]` to a function that expects `List[Tuple[str, Optional[str]]]`. When I tried changing the return type of `_parse_contributors()`, and the declared type of `maintainers`, to `List[Tuple[str, Optional[str]]]`, mypy passes locally. So in theory I think we can make that change along with requiring `pyproject-metadata>=0.8.1` and it should be enough...?

[the remark this answers is not in the document]
Comment (diazona) [reply test_support/test_support/metadata.py]: The other option is to dynamically check the version of pyproject-metadata available and switch the types if it's too old, but that's tricky and IMO not worth the trouble unless someone actually needs it.

[the remark this answers is not in the document]
Comment (sjlongland) [reply test_support/test_support/metadata.py]: Yeah, I really do not like overriding a check like this… but I think that's favourable to implementing more complex logic that dynamically switches data types.

[the remark this answers is not in the document]
Comment (diazona) [reply test_support/test_support/metadata.py]: Oh I think I'd actually prefer to dynamically switch, if we wanted to take that route, because that way we still get the benefits of type checking on `authors` and `maintainers`, as opposed to overriding the check with a comment (if that's what you meant) where we get no type checking at all. I don't even think it'd be that complex; we could just define a type alias differently depending on which version is available.

```python
if Version(importlib.metadata.version("pyproject-metadata")) >= Version("0.8.1"):
    Contributor = Tuple[str, Optional[str]]
else:
    Contributor = Tuple[str, str]
```

or something like that.

But in this case I'm banking on the fact that there's no particular reason to support older versions of pyproject-metadata, and thus no little or no incentive to make that type alias at all.

Although... if we're dropping support for `pyproject-metadata<0.8.1`, that might also affect which versions of Python we can run tests on. It looks like we'd be required to use Python 3.7... Maybe we could disable the tests in the `distribution/` directory on Python 3.6 and keep the rest, but if that's a hassle, I wouldn't mind dropping support for Python 3.6 entirely. It's old enough not to justify going to any trouble to keep it supported. (Ordinarily I would make the change of dropping a Python version in a separate PR, but if it has to happen, it might have to be combined with this one in order to pass CI.)

[the remark this answers is not in the document]
Comment (sjlongland) [reply test_support/test_support/metadata.py]: Yeah, Python 3.6 and the Linux distributions that require it are getting pretty long in the tooth now.

Probably old enough that we start sunsetting support for it, partial support in that we'll test basic functionality, but for reliable behaviour, we recommend something newer.

How difficult is it to exclude a particular test directory for a specific Python version?  If it's too much trouble, moving the minimum CI-tested release to 3.7 isn't a show-stopper.  We're a 2-person team not a multi-national software company. :-)

[the remark this answers is not in the document]
Comment (diazona) [reply test_support/test_support/metadata.py]: It depends on how you drastically you want to exclude the directory. E.g. if you just want to prevent the tests in a file from being run on certain Python versions, you can [use `pytest.mark.skipif()`](https://docs.pytest.org/en/latest/how-to/skipping.html#skip-all-test-functions-of-a-class-or-module), and that could be applied to every file in a directory either manually, or perhaps using `conftest.py`. If some code in a file won't even compile for a certain Python version, then it's necessary to [take more drastic measures to prevent it from even being read](https://docs.pytest.org/en/latest/how-to/skipping.html#skipping-files-or-directories).

In this case, though, the problem isn't really with running the tests, it's with static type checking. And our code has basically never passed static type checking on Python 3.6 (or wouldn't have, if we'd ever run it), and there's nothing we can do about that because there's no version of `pyproject-metadata` that supports Python 3.6 (which I had forgotten when I was writing my last comment). So I'd say we're clear to bump the minimum version of `pyproject-metadata` to 0.8.1 without any other metadata changes or having to exclude any tests.

I guess we've taken a long and winding road to arrive at this (at least this would be my recommendation, assuming I haven't missed anything):

- Change the type to `List[Tuple[str, Optional[str]]]` in those two places
- Change the dependency on `pyproject-metadata` to require version 0.8.1, minimum
- If that works, that's it

[the remark this answers is not in the document]
Comment (sjlongland) [reply test_support/test_support/metadata.py]: Right, so just having a look at this now… (it's been a busy few weeks, but I'm now working part-time, so I'll have a little more time for projects like this going forward)…

Looks like if we require `pyproject-metadata >= 0.8.1`, that is basically going to force our minimum Python version up from v3.6 to v3.7.

https://pypi.org/project/pyproject-metadata/0.8.1/

I think this is a fair compromise.  Python 3.7 is the standard version in Debian 1o (Buster).  Ubuntu 18.04 (Bionic) ships with 3.6, but has 3.7 available in `bionic-updates`.  Fedora 29 ships with 3.7.2.  None of these OSes have mainstream support (I think Ubuntu still provides support for 18.04, but you must pay for extended support for that).  On the commercial OS front, Windows 7 supports up to Python 3.8 and I think MacOS X 10.6 will also run 3.8~3.9 or so.

In short, most OSes that currently receive security fixes, and run a mainstream supported release of Python, would be excluded by moving the minimum to 3.7.

I think I'll bump `pyproject-metadata` and Python minimum versions then. :-)

**Comment by sjlongland** [comment]
…and joy of joys `pyproject-metadata > 0.8.1` breaks:

```
test_support/test_support/metadata.py:128: error: Item "RFC822Message" of "RFC822Message | Any" has no attribute "body"  [union-attr]
test_support/test_support/metadata.py:129: error: Item "RFC822Message" of "RFC822Message | Any" has no attribute "body"  [union-attr]
test_support/test_support/metadata.py:235: error: Item "RFC822Message" of "RFC822Message | Any" has no attribute "headers"  [union-attr]
```

It seems on the road to 0.9.0 earlier this year, they made `RFC822Message` a sub-class of `email.message.EmailMessage`, which now breaks our code.

**Comment by sjlongland** [comment]
Right… so after two bumps of `pyproject_metadata` and working through the breakage that this caused, I think I've now got it to some respectable shape.  Python 3.7 is now the minimum version required.

File: test_support/test_support/metadata.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (diazona) [reply test_support/test_support/metadata.py]: > Looks like if we require pyproject-metadata >= 0.8.1, that is basically going to force our minimum Python version up from v3.6 to v3.7.

It doesn't, though. We only use `pyproject-metadata` in test support code for tests that don't run on Python 3.6, so we can continue to support Python 3.6 regardless of what we do about the version of `pyproject-metadata`.

[the remark this answers is not in the document]
Comment (diazona) [reply test_support/test_support/metadata.py]: As further context: the reason I'd care about keeping support for Python 3.6 is not really about OS support, it's about supporting old corporate or academic environments that just haven't been updated in too long. Like, at my work, I think we were still using 3.6 in some capacity as recently as earlier this year (TBH we might _still_ be), not because the OS doesn't support it, but because we have a really small infrastructure support team and it took that long for somebody to get the time to do the upgrade and test it. It just removes one little barrier to migration for those people if we continue to support old versions.

That being said, of course we have to draw the line somewhere, and we'll surely wind up dropping support for 3.6 eventually. I'm just reluctant to do it because of something minor which is easily worked around. The argument I'm making here is essentially that the `pyproject-metadata` dependency is something minor which is easily worked around - specifically, we've already been working around it by excluding the distribution package tests on Python 3.6.

[the remark this answers is not in the document]
Comment (sjlongland) [reply test_support/test_support/metadata.py]: No problems, well I've re-arranged things to pull out the 3.7 minimum requirement out of this PR.

File: .github/workflows/docs.yml [hunk 4]
```diff
@@ -42,7 +42,9 @@ jobs:
...
         uses: actions/setup-python@v4
+      - name: Set up venv
+        run: python -m venv .pyenv
       - name: Install tox
-        run: python -m pip install tox
-      - name: Run
-        run: tox -e docs
+        run: .pyenv/bin/python -m pip install tox
+      - name: Run tox
+        run: .pyenv/bin/python -m tox -e docs
```
Comment (diazona) [inline .github/workflows/docs.yml:50]: OK so I'm back to looking at this after far too long, and to refresh my memory... if we didn't use a venv here, would we get the error about `--break-system-packages`? I mean, I'd prefer not to make this change without a reason, but if using the system Python would fail unless we use `--break-system-packages`, that's enough of a reason for me. :+1:

**Review by diazona — APPROVED** [review]
I finally got around to re-reviewing this and I think everything makes sense, except for that one question I had about using a venv - but that's not a big enough deal to be a blocker so I'll just approve.

File: .github/workflows/docs.yml (hunk 4 above)
Comment (sjlongland) [reply .github/workflows/docs.yml:50]: I think that was the case… it's been a while so my memory is fuzzy as well.

I could use `--break-system-packages`, as this is a container, a virtual environment in its own right, so nothing would break if we did use that flag (or tell `pip` to assume `--break-system-packages`), but I'm trying to break myself out of that habit. :-)

Comment (diazona) [reply .github/workflows/docs.yml:50]: Yeah that sounds like exactly what I was thinking. So we can just go ahead with this as is. I'll leave it to you to merge, in case there's anything else I missed.
